pynicotine/downloads.py

```diff
@@ -1046,21 +1052,26 @@ def retry_downloads(self, downloads):
             bypass_filter = (num_downloads == 1 and download.status == "Filtered")
             self.retry_download(download, bypass_filter)
 
-    def retry_download_limits(self):
+    def retry_limited_downloads(self, username):
 
         limited_statuses = {"Too many files", "Too many megabytes"}
 
         for download in self.transfers.values():
-            if download.status in limited_statuses or download.status.startswith("User limit of"):
-                # Re-queue limited downloads every 12 minutes
+            if download.username != username:
+                continue
 
-                log.add_transfer("Re-queuing file %(filename)s from user %(user)s in download queue", {
-                    "filename": download.virtual_path,
-                    "user": download.username
-                })
+            if download.status not in limited_statuses and not download.status.startswith("User limit of"):
+                continue
 
-                self.abort_download(download, abort_reason=None)
-                self.get_file(download.username, download.virtual_path, transfer=download)
+            log.add_transfer("Re-queuing file %(filename)s from user %(user)s in download queue", {
+                "filename": download.virtual_path,
+                "user": download.username
+            })
+
+            self.abort_download(download, abort_reason=None)
+            self.get_file(download.username, download.virtual_path, transfer=download)
+
+        del self.limited_user_timers[username]
```

Comment:
Consider `break` out of the loop to only re-queue a single file, and come back in another 5 minutes to do the next download. After the last set a flag for if not any download found in transfers then delete username timer.

It might help to avoid the peer banning us due to hitting the limit so hard.

Reply:
Only re-queuing a single file is a bit ridiculous when transfer speeds are fast, but we could limit the number of re-queued downloads to something reasonable, e.g. 100 (same as the default file limit for upload queues in Nicotine+).

Reply:
> we could limit the number of re-queued downloads to something reasonable, e.g. 100

If the limit set by the uploader could be determined with certainty, such as from knowing the number or total size of downloads that we had queued before encountering the limit, then it might be useful intelligence for the module to cache this to set the request rate deterministically, albeit this would be a lot more complex.